docs/proposals/virtualhost-policy.md

```diff
@@ -0,0 +1,98 @@
+# VirtualHost policy
+
+## Context
+
+Since Kuma 1.1.4 the dataplane now embeds a CoreDNS instance which forwards all DNS traffic to envoy proxy.
+This opens the door to more flexible implementation of vips.
+We propose here to create a new policy which will enable users to create arbitrary hostnames and ports that will route to a specific sets of tags.
+
+## New Policy
+
+We will add a new policy type:
+
+```yaml
+type: VirtualHost
+mesh: string # The mesh this policy applies to
+name: string #The name of the policy
+tags:
+    # A map of tags that should be taken for creating the tag selector. 
+    # It can either have a value or have "*" in which case a different vip will be generated for each tuple
+host: string # A go template with a special function `tag` which can use tags present in the `tags` section.
+port: int # A port to expose the virtual host on
+```
+
+It is impossible to create a VirtualHost that hasn't got a tag `kuma,io/service`.
+
+### Example
+
+Simple service with no wildcard replacement:
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service
+tags:
+  kuma.io/service: my-service
+host: httpbin.mesh
+port: 8080
+```
+
+Create different hostnames for each version of a service:
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service-by-version
+tags:
+  kuma.io/service: "*"
+  version: "*"
+host: "{{tag version}}.{{tag kuma.io/service}}.mesh"
+port: 8080
```

Comment:
we could also consider making a port a string and try to infer it from tag if needed, so

```
port: "{{tag kuma.io/virtual-host-port}}"
```

Reply:
I thought some more about this and it's actually tricky to make something not too error prone especially when you start having virtual hosts that cross multiple Dataplanes with many different tags.
I think it's also a feature that can be added later on if needed.